fix(ivy): wrong context passed to ngOnDestroy when resolved multiple times

crisbeto opened the pull request:
When the same provider is resolved multiple times on the same node, the first invocation had the correct context, but all subsequent ones were incorrect because we were registering the hook multiple times under different indexes in `destroyHooks`.

Fixes #35167.

Azbesciak commented on packages/core/src/render3/di_setup.ts:
```diff
@@ -81,26 +86,25 @@ function resolveProvider(
             tView, token);
+        if (providerIsTypeProvider || isClassProvider(provider)) {
+          const prototype = ((provider as ClassProvider).useClass || provider).prototype;
+          const ngOnDestroy = prototype.ngOnDestroy;
```
What with `useFactory` for example when you provide some object which is not a class instance?

crisbeto replied:
I'm not sure about that one, but I'll bring it up. The previous approach didn't support `useFactory` either.

Azbesciak commented:
@crisbeto yeah, you are right. That would also explain why I always call `ngOnDestroy` from my root components manually also... because I noticed that this is sometimes called, sometimes not.

waterplea commented on packages/core/src/render3/di_setup.ts:
```diff
@@ -17,10 +17,12 @@ import {DirectiveDef} from './interfaces/definition';
 
 
+/** Any provider definition that's not a `TypeProvider`. */
+type NonTypeProvider = Exclude<Exclude<Provider, TypeProvider>, any[]>;
```
Wouldn't it work with a single Exclude?
`Exclude<Provider, TypeProvider | any[]>`

johannesnormannjensen commented:
This fix is really needed

Azbesciak commented:
@pkozlowski-opensource could you... :)

crisbeto committed:
fix(ivy): wrong context passed to ngOnDestroy when resolved multiple times

When the same provider is resolved multiple times on the same node, the first invocation had the correct context, but all subsequent ones were incorrect because we were registering the hook multiple times under different indexes in `destroyHooks`.

Fixes #35167.

kara reviewed (approved):
LGTM

kara commented:
[presubmit](http://test/OCL:295839203:BASE:295839229:1582069177109:8308dc40)
